### PR TITLE
Change in astype leads to no longer casting to numpy string dtypes

### LINKED ISSUE #39566: BUG: Never end up with numpy string dtypes
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
DataFrame(["foo", "bar", "baz"]).astype(bytes)

```
As discussed in #39484 we do not want to get numpy string dtypes here. This should have dtype object

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #39566
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

@jreback as discussed we no longer cast to numpy string dtypes after this

### CODE REVIEW

**Comment by phofl** [comment]
I am not familiar with the experimental data manager. What could be the reason that the tests fail there but not on azure? Additionally only the DataFrame tests seem to fail, not the series tests

**Comment by jbrockmendel** [comment]
> What could be the reason that the tests fail there but not on azure?

they only are run if you specify add `--array-manger` to the pytest command, so only get run in that one build

**Comment by phofl** [comment]
Thx, this explains the behavior and the not raising in series methods.

@jorisvandenbossche You added a TODO about the numpy arrays in there. Did you have any concrete plans handling this case?

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -1268,7 +1268,7 @@ def soft_convert_objects(
             values = lib.maybe_convert_objects(
                 values, convert_datetime=datetime, convert_timedelta=timedelta
             )
-        except (OutOfBoundsDatetime, ValueError):
+        except OutOfBoundsDatetime:
```
Comment (jbrockmendel) [inline pandas/core/dtypes/cast.py:1271]: nice!

File: pandas/core/internals/array_manager.py [hunk 2]
```diff
@@ -287,7 +287,9 @@ def apply(
                 if not ignore_failures:
                     raise
                 continue
-            # if not isinstance(applied, ExtensionArray):
+            if not isinstance(applied, ExtensionArray):
+                if issubclass(applied.dtype.type, (str, bytes)):
+                    applied = np.array(applied, dtype=object)
```
Comment (jbrockmendel) [inline pandas/core/internals/array_manager.py:292]: i think we ought to have a sanitize_str_dtypes or something akin to sanitize_to_nanoseconds

**Review by jorisvandenbossche — COMMENTED** [review]
I think this will also need some more tests. Eg currently we also preserve the bytes dtype on setitem (`df["new_column"] == np.array(..)`), and not only for `astype`.  
In addition, we need to explicitly test the content of the resulting series after the astype operation (are the values converted to bytes?)

File: pandas/core/internals/array_manager.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/internals/array_manager.py:292]: This is now done as part of maybe_coerce_values, though that may be more heavy-weight than we want here

File: pandas/core/internals/array_manager.py [hunk 3]
```diff
@@ -413,7 +415,10 @@ def downcast(self) -> ArrayManager:
         return self.apply_with_block("downcast")
 
     def astype(self, dtype, copy: bool = False, errors: str = "raise") -> ArrayManager:
-        return self.apply("astype", dtype=dtype, copy=copy)  # , errors=errors)
+        # if issubclass(dtype, (str, bytes)):
+        #     dtype = "object"
+        y = self.apply("astype", dtype=dtype, copy=copy)  # , errors=errors)
+        return y
```
Comment (jbrockmendel) [inline pandas/core/internals/array_manager.py:421]: can the commented-out code be removed?

**Comment by simonjayhawkins** [comment]
@phofl can you merge master to resolve conflicts and address comments

**Comment by simonjayhawkins** [comment]
@phofl closing as stale. reopen when ready.

**Comment by phofl** [comment]
Yes will reopen when fixed, not much time right now unfortunately

**Pull request closed without merging**